### PR TITLE
Users should be able to search for covid testing centers.

### PR DESCRIPTION
**Pull request by azaria-dedmon**
**What does this PR do?**

- This PR provides user search functionality for covid testing centers.

**Description of task to be completed.**

- Implement homepage functionality and include state drop down list in `homepage.html`.
- Include covid testing API for testing locations in `init.py`.
- Include Geocode API for latitude and longitude in `init.py`.
- Include Mapbox API for map functionality in `location.html`.
- Include relevant unit tests.

**How should this be manually tested?**

- Run `Python3 run.py`
- Go to `localhost:5000/`
- Select a state from the drop down list.
- Click the search button.

### CODE REVIEW

**Comment by azaria-dedmon** [comment]
@alexkayabula I've found what the issue is. Printing the code like this, "console.log(**{{coors}}**) was causing JS to execute the console.log statement with double parentheses. Since the latitude and longitude were paired in a tuple, the parentheses were already present and the console.log's parentheses were not needed. As a result, I got rid of the console.log(**{{coors}}**) and just did console.log{{cors}}. The pair is showing up now in JS.

**Comment by alexkayabula** [comment]
Thats great. I actually didn't realize that.

**Comment by azaria-dedmon** [comment]
@alexkayabula Hey Alex. Im having issues with my tests passing. I keep receiving this error from  travis ,
```
File "/home/travis/build/azaria-dedmon/covid-19/app/refactor.py", line 2, in <module>
    from .privateinfo import key
ModuleNotFoundError: No module named 'app.privateinfo
```
To my knowledge I am importing the API key correctly in the refactor file, but things are still not working out properly.

**Comment by alexkayabula** [comment]
> @alexkayabula Hey Alex. Im having issues with my tests passing. I keep receiving this error from travis ,
> 
> ```
> File "/home/travis/build/azaria-dedmon/covid-19/app/refactor.py", line 2, in <module>
>     from .privateinfo import key
> ModuleNotFoundError: No module named 'app.privateinfo
> ```
> 
> To my knowledge I am importing the API key correctly in the refactor file, but things are still not working out properly.

Let me check it out.

**Comment by azaria-dedmon** [comment]
@alexkayabula Hey so I made some changes and decided to hide the API key using an environment variable instead of importing it from an ignored file. However, my unit tests are failing now. I believe it has something to do with the environment variable. I'm looking into it right now.

**Comment by azaria-dedmon** [comment]
@alexkayabula Would it be possible that hiding my API key using an environment variable could cause the test that relies on that key to fail? When the API key was hard-coded in my main file, all of the unit tests were passing, but now that I am using an OS variable, its not anymore.

**Comment by alexkayabula** [comment]
> @alexkayabula Hey so I made some changes and decided to hide the API key using an environment variable instead of importing it from an ignored file. However, my unit tests are failing now. I believe it has something to do with the environment variable. I'm looking into it right now.

@azaria-dedmon I think the reason why the tests are failing is because you are trying to import from an ignored file. What is contained in the `privateinfo file`?

**Comment by azaria-dedmon** [comment]
@alexkayabula The privateinfo file contained the API key. After trial and error I decided to stop using that file and just use an environment variable for the API key. But the tests are still failing

**Comment by alexkayabula** [comment]
> @alexkayabula Would it be possible that hiding my API key using an environment variable could cause the test that relies on that key to fail? When the API key was hard-coded in my main file, all of the unit tests were passing, but now that I am using an OS variable, its not anymore.

Keep the API as an environment variable and set it in Travis CI. [https://docs.travis-ci.com/user/environment-variables/#defining-variables-in-repository-settings](https://docs.travis-ci.com/user/environment-variables/#defining-variables-in-repository-settings)

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 venv/
 __pycache__/
 *.pyc
+privateinfo.py
```
Comment (alexkayabula) [inline .gitignore]: Do away with this line since you are not using this file.

File: app/refactor.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import os 
+import requests
+
+
+def test_code(state, API_BASE_URL):
+
+    key = os.environ.get('key')
```
Comment (alexkayabula) [inline app/refactor.py]: Keep this as an environment variable and set the key in Travis Ci.

**Comment by azaria-dedmon** [comment]
> > @alexkayabula Would it be possible that hiding my API key using an environment variable could cause the test that relies on that key to fail? When the API key was hard-coded in my main file, all of the unit tests were passing, but now that I am using an OS variable, its not anymore.
> 
> Keep the API as an environment variable and set it in Travis CI. https://docs.travis-ci.com/user/environment-variables/#defining-variables-in-repository-settings

Thank you👍🏾

File: app/__init__.py [hunk 3]
```diff
@@ -1,13 +1,15 @@
 """Application."""
-
+import os
+import requests
```
Comment (alexkayabula) [inline app/__init__.py:3]: @azaria-dedmon save this to the requirements.txt file.

File: app/__init__.py [hunk 4]
```diff
@@ -88,10 +94,19 @@ def logout():
     """Handle user logout."""
     do_logout()
     flash("Success!", "success")
-    return redirect("/")
+    return redirect('/')
```
Comment (alexkayabula) [inline app/__init__.py]: @azaria-dedmon this change is not related to searching for covid centers.

File: app/templates/location.html [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            {% endif %}
+            <li><a href="/">Home</a></li>
+          </ul>
+    </div>
+    </div>
+    <div id='map'></div>
+
+    <script>
+
+        mapboxgl.accessToken = 'pk.eyJ1IjoiYXphcmlhZGVkbW9uIiwiYSI6ImNrMmxmdjgwZjA2b3kzcHQ1dWh5Znc1NDcifQ.2VHaTw15QaiOknOoW8aP4A';
```
Comment (alexkayabula) [inline app/templates/location.html:31]: @azaria-dedmon It is not good practice to have this exposed in the code.

**Review by alexkayabula — CHANGES_REQUESTED** [review]

**Comment by alexkayabula** [comment]
@azaria-dedmon Basing on the use of the `refactor.py`, Is it possible to rename that file to something else ? I find that name abit confusing. What do you think ?

**Comment by azaria-dedmon** [comment]
> @azaria-dedmon Basing on the use of the `refactor.py`, Is it possible to rename that file to something else ? I find that name abit confusing. What do you think ?

I figured `refactor` would be a good name for that file since the code required for that route seemed a bit too large for the endpoint and needed to be moved to another file for readability. If it is confusing/misleading, I will go ahead and rename it.

**Comment by alexkayabula** [comment]
> > @azaria-dedmon Basing on the use of the `refactor.py`, Is it possible to rename that file to something else ? I find that name abit confusing. What do you think ?
> 
> I figured `refactor` would be a good name for that file since the code required for that route seemed a bit too large for the endpoint and needed to be moved to another file for readability. If it is confusing/misleading, I will go ahead and rename it.

Rename it basing on its functionality.

**Comment by alexkayabula** [comment]
@azaria-dedmon in the `How should this be manually tested?` part of the description, you should add the click search button step.

**Comment by alexkayabula** [comment]
@azaria-dedmon what is the main functionality of the `test_code` function in the `state_testing_code`file ?

**Comment by azaria-dedmon** [comment]
> @azaria-dedmon what is the main functionality of the `test_code` function in the `state_testing_code`file ?

To search for testing centers.

**Comment by alexkayabula** [comment]
> > @azaria-dedmon what is the main functionality of the `test_code` function in the `state_testing_code`file ?
> 
> To search for testing centers.

OK, basing on the implementation, the `test_code` function returns the coordinate

> > @azaria-dedmon what is the main functionality of the `test_code` function in the `state_testing_code`file ?
> 
> To search for testing centers.

How about `latsLngs` ? Does it represent/contain only the coordinates or the location details i.e phone, place, latitude, longitude etc

File: app/state_testing_code.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import os
+import requests
+
+def test_code(state, API_BASE_URL):
```
Comment (alexkayabula) [inline app/state_testing_code.py]: Since this function actually returns/gets the testing locations in a given state, we should rename it to `get_testing_locations`.

File: app/state_testing_code.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+    key = os.environ.get('key')
+
+    url = f'https://covid-19-testing.github.io/locations/{state.lower()}/complete.json'
+    res = requests.get(url)
+    testing_data = res.json()
+    latsLngs = {}
```
Comment (alexkayabula) [inline app/state_testing_code.py]: Since this variable contains the actual testing locations in a state, we can rename it and corresponding occurences to `locations`

File: app/__init__.py [hunk 8]
```diff
@@ -94,4 +100,13 @@ def logout():
...
 def show_homepage():
     """Shows website homepage"""
-    return render_template('homepage.html')
\ No newline at end of file
+    return render_template('homepage.html', testing_states=testing_states)
+
+
+@app.route('/location')
+def show_state_locations():
+    """Return selected state from drop down menu"""
```
Comment (alexkayabula) [inline app/__init__.py]: Since this function shows the testing locations in a state, we can restructure the comment to `Return testing locations for state selected from drop down menu`

File: app/__init__.py [hunk 9]
```diff
@@ -94,4 +100,13 @@ def logout():
...
+    state = request.args.get('state')
+
+    latsLngs = test_code(state, API_BASE_URL)
```
Comment (alexkayabula) [inline app/__init__.py]: Rename variable and function accordingly.

File: app/__init__.py [hunk 10]
```diff
@@ -1,13 +1,15 @@
...
-
+import os
+import requests
 from flask import Flask, session, g, render_template, redirect, request, flash
 from forms import RegisterUser, LoginUser
-from models import connect_db, User, db
+from models import connect_db, User, db, testing_states
 from config import config
 from sqlalchemy.exc import IntegrityError
+from .state_testing_code import test_code
```
Comment (alexkayabula) [inline app/__init__.py]: Refactor as instructed.

**Review by alexkayabula — CHANGES_REQUESTED** [review]

**Comment by alexkayabula** [comment]
@azaria-dedmon you should also rename the file to `locations_helper.py`.

**Comment by azaria-dedmon** [comment]
> > > @azaria-dedmon what is the main functionality of the `test_code` function in the `state_testing_code`file ?
> > 
> > 
> > To search for testing centers.
> 
> OK, basing on the implementation, the `test_code` function returns the coordinate
> 
> > > @azaria-dedmon what is the main functionality of the `test_code` function in the `state_testing_code`file ?
> > 
> > 
> > To search for testing centers.
> 
> How about `latsLngs` ? Does it represent/contain only the coordinates or the location details i.e phone, place, latitude, longitude etc

It represents the location details. Considering that I am passing this info into the features array within the `geojson` variable, I think `location-features` could be fitting.

**Comment by alexkayabula** [comment]
> > > > @azaria-dedmon what is the main functionality of the `test_code` function in the `state_testing_code`file ?
> > > 
> > > 
> > > To search for testing centers.
> > 
> > 
> > OK, basing on the implementation, the `test_code` function returns the coordinate
> > > > @azaria-dedmon what is the main functionality of the `test_code` function in the `state_testing_code`file ?
> > > 
> > > 
> > > To search for testing centers.
> > 
> > 
> > How about `latsLngs` ? Does it represent/contain only the coordinates or the location details i.e phone, place, latitude, longitude etc
> 
> It represents the location details. Considering that I am passing this info into the features array within the `geojson` variable, I think `location-features` could be fitting.

Maybe `location_details` since it actually contains details of different testing locations.

**Comment by alexkayabula** [comment]
@azaria-dedmon Kindly review and fix the requested changes so that we get this merged ASAP.

**Comment by alexkayabula** [comment]
Great work on this on.
